satpy/tests/test_multiscene.py

```diff
@@ -211,80 +213,96 @@ def test_from_files(self):
-        from satpy.multiscene import add_group_aliases
+    def test_group_alias_equals_original_dataset(self, grouped_multi_scene, scene1):
+        """Test that the group alias equals the original dataset."""
+        scene1_grp = grouped_multi_scene.scenes[0]
+        xr.testing.assert_allclose(scene1_grp["ds1"], scene1["ds1"])
 
-        # Define test scenes
-        ds_id1 = make_dataid(name='ds1', wavelength=(10.7, 10.8, 10.9))
-        ds_id2 = make_dataid(name='ds2', wavelength=(1.9, 2.0, 2.1))
-        ds_id3 = make_dataid(name='ds3', wavelength=(10.8, 10.9, 11.0))
-        ds_id31 = make_dataid(name='ds31', polarization='H')
+    def test_grouping_does_not_modify_orignal_dataset(self, grouped_multi_scene, scene1):
+        """Test that grouping does not modify the original dataset."""
+        assert DataQuery(name='odd') not in scene1
 
-        scene1 = Scene()
-        scene1[ds_id1] = xr.DataArray([1])
-        scene2 = Scene()
-        scene2[ds_id2] = xr.DataArray([2])
-        scene3 = Scene()
-        scene3[ds_id3] = xr.DataArray([3])
-        scene3[ds_id31] = xr.DataArray([4])
-        scenes = [scene1, scene2, scene3]
-
-        # Define groups
-        g1 = make_dataid(name='g1', wavelength=(10, 11, 12))
-        g2 = make_dataid(name='g2', wavelength=(1, 2, 3), polarization='V')
-        groups = {g1: ['ds1', 'ds3'], g2: ['ds2']}
-
-        # Test adding aliases
-        with_aliases = add_group_aliases(iter(scenes), groups)
-        self.assertIsInstance(with_aliases, types.GeneratorType)
-        with_aliases = list(with_aliases)
-        self.assertSetEqual(set(with_aliases[0].keys()), {g1, ds_id1})
-        self.assertSetEqual(set(with_aliases[1].keys()), {g2, ds_id2})
-        self.assertSetEqual(set(with_aliases[2].keys()), {g1, ds_id3, ds_id31})
-
-        np.testing.assert_array_equal(with_aliases[0]['g1'].values, [1])
-        np.testing.assert_array_equal(with_aliases[0]['ds1'].values, [1])
-        np.testing.assert_array_equal(with_aliases[1]['g2'].values, [2])
-        np.testing.assert_array_equal(with_aliases[1]['ds2'].values, [2])
-        np.testing.assert_array_equal(with_aliases[2]['g1'].values, [3])
-        np.testing.assert_array_equal(with_aliases[2]['ds3'].values, [3])
-        np.testing.assert_array_equal(with_aliases[2]['ds31'].values, [4])
-
-        # Make sure that modifying the result doesn't modify the original
-        self.assertNotIn(g1, scene1)
-
-        # Adding an alias for multiple datasets in one scene should fail
-        gen = add_group_aliases([scene3], {g1: ['ds3', 'ds31']})
-        self.assertRaises(ValueError, list, gen)
+    def test_fails_to_add_multiple_datasets_from_the_same_scene_to_a_group(self, multi_scene):
+        """Test that multiple datasets from the same scene in one group fails."""
+        groups = {DataQuery(name='mygroup'): ['ds1', 'ds2']}
+        multi_scene.group(groups)
+        with pytest.raises(ValueError):
+            next(multi_scene.scenes)
```

Comment:
I'm not sure how I feel about this way of testing. I'm curious what @mraspaud and @pnuu think. From what I understand of it you've made a test for each *check* but they use a single test case essentially via the `grouped_multi_scene` fixture. I think my preference would be to have all the checks in a single test method, but I can be persuaded to be OK with this.

Reply:
> I tried to follow the recommendation of making only one assertion per test.

Is that a "Clean Code" recommendation?

Reply:
I think so. Dave Farley mentioned it in a Continuous Delivery Video.

Reply:
Something along the lines "A test should fail for only one reason"

Reply:
`<pulls "Clean Code" from bookshelf next to desk>`

The testing examples in Clean Code by Robert C Martin have multiple asserts in a single test.

Reply:
If you can find the Dave Farley video that'd be good. Actually...I have Dave Farley's book downstairs, let me see if he says anything in there. To me failing for one reason could mean something different than "there is only one spot in the test that can fail"

Reply:
Found it!

https://youtu.be/UWtEVKVPBQ0?t=584

He does seem to be saying exactly what you've done here, but I'm not sure I agree. We may want to discuss this with others. To me there is a fine line about what he is saying and *maybe* what he actually means. I get what he's saying about having multiple asserts but I would like to think the main takeaway is that you are only checking *one* behavior not necessarily checking only one assertion. For example,

```
def test_data_copy():
    orig_data = ...
    copied_data = orig_data.copy()
    copied_data["new_key"] = "new_value"
    assert copied_data["new_key"] == "new_value"
    assert "new_key" not in copied_data
```

This test contains two assertions, but it is really only testing one overall behavior. ...but now that I'm writing this I see how the first assertion is testing the "behavior" that setting data worked and the second assertion is testing that the copy worked. But I suppose the setting would be yet another test while the setting here is just one necessity of testing the copy. :man_shrugging:

Reply:
I agree that the definition of behaviour is a bit vague. I just tried it out to see how it feels. And for me the fact that I had to come up with a very explicit name for those tiny tests made the biggest difference in readability.

Reply:
So I simplified the tests as we discussed in the meeting, but now I feel an urgent need to explain the intent of these assertions with comments...